File: library/src/main/java/me/proxer/library/api/user/UserApi.java
package me.proxer.library.api.user;

import lombok.experimental.Accessors;
import retrofit2.Retrofit;

import javax.annotation.Nullable;

/**
 * API for the User class.
 *
 * @author Ruben Gees
 */
@Accessors(fluent = true)
public final class UserApi {

    private final InternalApi internalApi;

    /**
     * Only for internal use.
     */
    public UserApi(final Retrofit retrofit) {
        this.internalApi = retrofit.create(InternalApi.class);
    }

    /**
     * Returns the respective endpoint.
     */
    public LoginEndpoint login(final String username, final String password) {
        return new LoginEndpoint(internalApi, username, password);
    }

    /**
     * Returns the respective endpoint.
     */
    public LogoutEndpoint logout() {
        return new LogoutEndpoint(internalApi);
    }

    /**
     * Returns the respective endpoint.
     */
    public UserTopTenEndpoint topTen(@Nullable final String userId, @Nullable final String username) {
        return new UserTopTenEndpoint(internalApi, userId, username);
    }

    /**
     * Returns the respective endpoint.
     */
    public UserInfoEndpoint info(@Nullable final String userId, @Nullable final String username) {
        return new UserInfoEndpoint(internalApi, userId, username);
    }

    /**
     * Returns the respective endpoint.
     */
    public UserAboutEndpoint about(@Nullable final String userId, @Nullable final String username) {
        return new UserAboutEndpoint(internalApi, userId, username);
    }

    /**
     * Returns the respective endpoint.
     */
    public UserMediaListEndpoint mediaList(@Nullable final String userId, @Nullable final String username) {
        return new UserMediaListEndpoint(internalApi, userId, username);
    }

    /**
     * Returns the respective endpoint.
     */
    public UserCommentsEndpoint comments(@Nullable final String userId, @Nullable final String username) {
        return new UserCommentsEndpoint(internalApi, userId, username);
    }

    /**
     * Returns the respective endpoint.
     */
    public UserHistoryEndpoint history(@Nullable final String userId, @Nullable final String username) {
        return new UserHistoryEndpoint(internalApi, userId, username);
    }
}

File: library/src/main/kotlin/me/proxer/library/api/user/UserApi.kt
package me.proxer.library.api.user

import retrofit2.Retrofit

/**
 * API for the User class.
 *
 * @author Ruben Gees
 */
class UserApi internal constructor(retrofit: Retrofit) {

    private val internalApi = retrofit.create(InternalApi::class.java)

    /**
     * Returns the respective endpoint.
     */
    fun login(username: String, password: String): LoginEndpoint {
        return LoginEndpoint(internalApi, username, password)
    }

    /**
     * Returns the respective endpoint.
     */
    fun logout(): LogoutEndpoint {
        return LogoutEndpoint(internalApi)
    }

    /**
     * Returns the respective endpoint.
     */
    fun topTen(userId: String? = null, username: String? = null): UserTopTenEndpoint {
        return UserTopTenEndpoint(internalApi, userId, username)
    }

    /**
     * Returns the respective endpoint.
     */
    fun info(userId: String? = null, username: String? = null): UserInfoEndpoint {
        return UserInfoEndpoint(internalApi, userId, username)
    }

    /**
     * Returns the respective endpoint.
     */
    fun about(userId: String? = null, username: String? = null): UserAboutEndpoint {
        return UserAboutEndpoint(internalApi, userId, username)
    }

    /**
     * Returns the respective endpoint.
     */
    fun mediaList(userId: String? = null, username: String? = null): UserMediaListEndpoint {
        return UserMediaListEndpoint(internalApi, userId, username)
    }

    /**
     * Returns the respective endpoint.
     */
    fun comments(userId: String? = null, username: String? = null): UserCommentsEndpoint {
        return UserCommentsEndpoint(internalApi, userId, username)
    }

    /**
     * Returns the respective endpoint.
     */
    fun history(userId: String? = null, username: String? = null): UserHistoryEndpoint {
        return UserHistoryEndpoint(internalApi, userId, username)
    }
}
